quantum/process_keycode/process_tap_dance.h

```diff
@@ -22,30 +22,27 @@
 #    include <inttypes.h>
 
 typedef struct {
+    uint16_t timer;
+    uint16_t interrupting_keycode;
     uint8_t  count;
     uint8_t  oneshot_mods;
     uint8_t  weak_mods;
-    uint16_t keycode;
-    uint16_t interrupting_keycode;
-    uint16_t timer;
-    bool     interrupted;
-    bool     pressed;
-    bool     finished;
+    bool     pressed : 1;
+    bool     finished : 1;
+    bool     interrupted : 1;
 } qk_tap_dance_state_t;
 
-#    define TD(n) (QK_TAP_DANCE | ((n)&0xFF))
-
 typedef void (*qk_tap_dance_user_fn_t)(qk_tap_dance_state_t *state, void *user_data);
 
 typedef struct {
+    qk_tap_dance_state_t state;
     struct {
         qk_tap_dance_user_fn_t on_each_tap;
         qk_tap_dance_user_fn_t on_dance_finished;
         qk_tap_dance_user_fn_t on_reset;
     } fn;
-    qk_tap_dance_state_t state;
-    uint16_t             custom_tapping_term;
-    void *               user_data;
+    uint16_t custom_tapping_term;
```

Comment:
This could probably be nuked to, in favor of the get_tapping_term support. 
```suggestion
```

Reply:
Yeah, that would clean up the tap-dance implementation nicely. I added a commit where I do so. I looked into usage of `ACTION_TAP_DANCE_FN_ADVANCED_TIME` and provided an alternative for all instances. It looks like some could just as well set `TAPPING_TERM` to the desired value, but I didn't want to be too invasive.

While this achieves a unified way of setting per-key tapping terms, it is somewhat more verbose for the keymaps.

If we want this, I assume we need a changelog entry, which I did not write.

Reply:
`ACTION_TAP_DANCE_FN_ADVANCED_TIME()`, which uses that field, was deprecated back in #6259 (merged on 25 Jul 2020) in favor of the generic custom tapping term support; that's probably enough time to get it removed completely.

Reply:
I'm a bit unsure what to do about the `TAPPING_TERM_PER_KEY` macro. It doesn't seem to be orthogonal to the other config settings such as `DYNAMIC_TAPPING_TERM` and `NO_ACTION_TAPPING`. Especially the latter I don't fully understand the intent of. If tap-dance is supposed to work even with `NO_ACTION_TAPPING`, we might want to have `#define get_tapping_term(keycode, event) (TAPPING_TERM)` somewhere on a `NO_ACTION_TAPPING` code path (this does mean that all the `get_tapping_term` definitions need to be guarded). Part of my confusion is that I don't really see how the setting of `TAPPING_TERM_PER_KEY` would have a big impact on the firmware size/speed.

Reply:
To keep this MR a little more focused, I restored the old `TAPPING_TERM_PER_KEY` behavior and spun out my thoughts on it into #16472. That should make this MR a bit easier to process.

Reply:
I rebased this MR on top of #16681. The churn is high, but that is mostly due to the last commit that drops `ACTION_TAP_DANCE_FN_ADVANCED_TIME` (as requested in this thread) and updates all the places in the repo where it was used.

As far as I can tell, this is ready to go in.

Reply:
This MR has been in development for 3 months now and the churn is only getting higher. @drashna, can you confirm we are still progressing toward an eventual merge? We now have a restructuring of the code, including more efficient action/state structs and simplification of the overall implementation, fixes to corner cases and initial test coverage, and deprecation of `ACTION_TAP_DANCE_FN_ADVANCED_TIME`.